src/Microsoft.OData.Client/BulkUpdateGraph.cs

```diff
@@ -0,0 +1,146 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="BulkUpdateGraph.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.OData.Client
+{
+    /// <summary>
+    /// Contains information about descriptors and their related descriptors. 
+    /// </summary>
+    internal sealed class BulkUpdateGraph
+    {
+        /// <summary>Set of related descriptors</summary>
+        private Dictionary<Descriptor, HashSet<Descriptor>> descriptorGraph;
+
+        /// <summary>Set of top level descriptors</summary>
+        private List<EntityDescriptor> topLevelDescriptors;
+
+        /// <summary> The EntitySetName of the top-level descriptor types </summary>
+        private string entitySetName;
+
+        /// <summary>where to pull the changes from</summary>
+        protected readonly RequestInfo requestInfo;
+
+        public BulkUpdateGraph(RequestInfo requestInfo)
+        { 
+            this.descriptorGraph= new Dictionary<Descriptor, HashSet<Descriptor>>(); 
+            this.topLevelDescriptors= new List<EntityDescriptor>();
+            this.requestInfo = requestInfo;
+        }
+
+        /// <summary>
+        /// Returns a collection of all the root descriptors used in a deep update.
+        /// </summary>
+        public List<EntityDescriptor> TopLevelDescriptors
+        {
+            get
+            {
+                return this.topLevelDescriptors;
+            }
+        }
+
+        /// <summary>
+        /// Returns the EntitySetName.
+        /// </summary>
+        public string EntitySetName
+        {
+            get
+            {
+                return this.entitySetName;
+            }
+        }
+
+        /// <summary>
+        /// Computes the EntyitySetName in cases where the top-level EntityDescriptors
+        /// do not have the EntitySetName set. 
+        /// </summary>
+        public void ComputeEntitySetName()
+        {
+            EntityDescriptor parentDescriptor = this.TopLevelDescriptors[0] as EntityDescriptor;

```

Comment:
[0] is safety?

Reply:
we do a check for the TopLevelDescriptors before calling this method.